compiler/src/iree/compiler/DispatchCreation/FormSplitReductionDispatches.cpp

```diff
@@ -27,6 +27,116 @@ namespace mlir::iree_compiler::DispatchCreation {
 
 namespace {
 
+/// Pattern to fold extract_slice(broadcast) into the broadcast input when
+/// the extract_slice is rank-reducing and only extracts along the broadcasted
+/// dimensions (with size 1), leaving all non-broadcasted dimensions intact.
+/// This is valid because the broadcast only duplicates data along the
+/// broadcasted dimension, and extracting a single slice from that dimension
+/// gives back the original input.
+///
+/// This fold is critical for correctness in parallel contexts: when the
+/// broadcast's outs operand comes from extracting a slice of a shared_outs
+/// tensor in scf.forall, the broadcast operation writes to that extracted
+/// slice, which aliases back to the shared tensor. Since the extracted slice
+/// spans parallel dimensions (e.g., [4, 1, 1] where dim 0 is the batch
+/// dimension), multiple parallel workgroups would all write to the same
+/// aliased memory location, causing a race condition. By folding to the
+/// broadcast input (which doesn't alias the shared tensor), subsequent passes
+/// can properly tile and distribute the initialization per workgroup, avoiding
+/// the race.
+///
+/// Example:
+///   %broadcast = linalg.broadcast ins(%in : tensor<4x1xf16>)
+///                                 outs(%out : tensor<4x1x1xf16>) dimensions =
+///                                 [2]
+///   %extract = tensor.extract_slice %broadcast[0, 0, 0] [4, 1, 1] [1, 1, 1]
+///              : tensor<4x1x1xf16> to tensor<4x1xf16>
+///   // Extracts all of dims 0,1 (sizes 4,1) and reduces dim 2 (size 1).
+/// ->
+///   %extract is replaced by %in (tensor<4x1xf16>)
+struct FoldExtractSliceOfBroadcast final
+    : OpRewritePattern<tensor::ExtractSliceOp> {
+  using Base::Base;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp extractOp,
+                                PatternRewriter &rewriter) const override {
+    auto broadcastOp =
+        extractOp.getSource().getDefiningOp<linalg::BroadcastOp>();
+    if (!broadcastOp) {
+      return failure();
```

Comment:
Nit: It is useful to add a `rewriter.notifyMatchFailure()` here with a line that says why the match failed. That will be printed out in the debug log.